snorkel/labeling/analysis.py

```diff
@@ -4,399 +4,349 @@
 import numpy as np
 import scipy.sparse as sparse
 from pandas import DataFrame, Series
+from sklearn.metrics import confusion_matrix
 
-from snorkel.analysis.error_analysis import confusion_matrix
 from snorkel.analysis.utils import to_int_label_array
 
-Matrix = Union[np.ndarray, sparse.csr_matrix]
 
-
-def _covered_data_points(L: Matrix) -> np.ndarray:
-    """Get indicator vector z where z_i = 1 if x_i is labeled by at least one LF."""
-    return np.ravel(np.where(L.sum(axis=1) != 0, 1, 0))
-
-
-def _overlapped_data_points(L: Matrix) -> np.ndarray:
-    """Get indicator vector z where z_i = 1 if x_i is labeled by more than one LF."""
-    return np.where(np.ravel((L != 0).sum(axis=1)) > 1, 1, 0)
-
-
-def _conflicted_data_points(L: sparse.spmatrix) -> np.ndarray:
-    """Get indicator vector z where z_i = 1 if x_i is labeled differently by two LFs."""
-    m = sparse.diags(np.ravel(L.max(axis=1).todense()))
-    return np.ravel(np.max(m @ (L != 0) != L, axis=1).astype(int).todense())
-
-
-def label_coverage(L: Matrix) -> float:
-    """Compute the fraction of data points with at least one label.
-
-    Parameters
-    ----------
-    L
-        Matrix where L_{i,j} is the label given by the jth LF to the ith candidate
-
-    Returns
-    -------
-    float
-        Fraction of data points with labels
-
-    Example
-    -------
-    >>> L = np.array([
-    ...     [0, 1, 1],
-    ...     [0, 0, 0],
-    ...     [2, 1, 0],
-    ...     [0, 1, 0],
-    ...     [1, 1, 1],
-    ... ])
-    >>> L = sparse.csr_matrix(L)
-    >>> label_coverage(L)
-    0.8
-    """
-    return _covered_data_points(L).sum() / L.shape[0]
-
-
-def label_overlap(L: Matrix) -> float:
-    """Compute the fraction of data points with at least two (non-abstain) labels.
-
-    Parameters
-    ----------
-    L
-        Matrix where L_{i,j} is the label given by the jth LF to the ith candidate
-
-    Returns
-    -------
-    float
-        Fraction of data points with overlapping labels
-
-    Example
-    -------
-    >>> L = np.array([
-    ...     [0, 1, 1],
-    ...     [0, 0, 0],
-    ...     [2, 1, 0],
-    ...     [0, 1, 0],
-    ...     [1, 1, 1],
-    ... ])
-    >>> L = sparse.csr_matrix(L)
-    >>> label_overlap(L)
-    0.6
-    """
-    return _overlapped_data_points(L).sum() / L.shape[0]
-
-
-def label_conflict(L: sparse.spmatrix) -> float:
-    """Compute the fraction of data points with conflicting (non-abstain) labels.
+class LFAnalysis:
+    """Run analyses on LFs using label matrix.
```

Comment:
Add Attributes

Reply:
So some of the methods use L (ABSTAIN = 0) and some use _L_sparse (ABSTAIN = -1). Is that just for convenience, or are there some of these helpers that absolutely require the original L in order to be calculated?

Reply:
Some only operate over NumPy arrays